File: src/main/go/leetcode/lists/hot100/LeetCode15.go
package hot100

import "sort"

func threeSum(nums []int) [][]int {
	if len(nums) == 0 {
		return [][]int{}
	}
	sort.Ints(nums)
	var res [][]int
	for i := 0; i < len(nums); i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		left, right := i+1, len(nums)-1
		for left < right {
			if nums[i]+nums[left]+nums[right] == 0 {
				res = append(res, []int{nums[i], nums[left], nums[right]})
				for left < right && nums[left+1] == nums[left] {
					left++
				}
				left++
				for left < right && nums[right-1] == nums[right] {
					right--
				}
				right--
			} else if nums[i]+nums[left]+nums[right] < 0 {
				left++
			} else {
				right--
			}
		}
	}
	return res
}

File: src/main/java/leetcode/lists/hot100/LeetCode15.java
package leetcode.lists.hot100;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/14 20:24
 */
public class LeetCode15 {

    //[-1,-1,0,1]
    //[0,1,1]
    //[0,0,0]

    public static void main(String[] args) {
        System.out.println(new LeetCode15().threeSum(new int[]{-1, -1, 0, 1}));
        System.out.println(new LeetCode15().threeSum(new int[]{0, 1, 1}));
        System.out.println(new LeetCode15().threeSum(new int[]{0, 0, 0}));
        System.out.println(new LeetCode15().threeSum(new int[]{0, 0, 0, 0}));
    }

    public List<List<Integer>> threeSum(int[] nums) {
        if (nums == null || nums.length == 0) {
            return new ArrayList<>();
        }

        Arrays.sort(nums);
        List<List<Integer>> list = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
            while (i >= 1 && i < nums.length && nums[i] == nums[i - 1]) {
                i++;
            }
            int left = i + 1;
            int right = nums.length - 1;
            while (left < right) {
                List<Integer> temp = new ArrayList<>();
                if (nums[left] + nums[right] + nums[i] == 0) {
                    temp.add(nums[i]);
                    temp.add(nums[left]);
                    temp.add(nums[right]);

                    list.add(temp);
                    while (left < nums.length - 1 && nums[left] == nums[left + 1]) {
                        left++;
                    }
                    while (right > 0 && nums[right] == nums[right - 1]) {
                        right--;
                    }
                    left++;
                    right--;
                } else if (nums[left] + nums[right] + nums[i] < 0) {
                    left++;
                } else {
                    right--;
                }
            }
        }
        return list;
    }
}
